internal/inputs/inputs.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package inputs
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/google/gke-policy-automation/internal/log"
+)
+
+const (
+	defaultMaxDataGetCoroutines = 20
+)
+
+type Input interface {
+	GetID() string
+	GetDescription() string
+	GetData(clusterID string) (interface{}, error)
+	Close() error
+}
+
+type Cluster struct {
+	Data map[string]interface{}
+}
+
+type getDataTask struct {
+	input     Input
+	clusterID string
+}
+
+type getDataTaskResult struct {
+	clusterID string
+	inputID   string
+	result    interface{}
+	err       error
+}
+
+//GetAllInputsData fetches data from given inputs for all given clusters in a concurrent manner
+func GetAllInputsData(inputs []Input, clusterIDs []string) (map[string]*Cluster, []error) {
+	tasksNo := len(inputs) * len(clusterIDs)
+	tasksChan := make(chan *getDataTask, tasksNo)
+	resultsChan := make(chan *getDataTaskResult, tasksNo)
+	errorsChan := make(chan *getDataTaskResult, tasksNo)
+	log.Infof("creating %d get data tasks", tasksNo)
+
+	for _, input := range inputs {
+		for _, clusterID := range clusterIDs {
+			tasksChan <- &getDataTask{input: input, clusterID: clusterID}
+		}
+	}
+	close(tasksChan)
+
+	log.Debugf("starting %d goroutines", defaultMaxDataGetCoroutines)
+	var wg sync.WaitGroup
```

Comment:
this processing can be done in one wrapping goroutine, to allow result and error channel to be significatly smaller (eq. `defaultMaxDataGetCoroutines`)
E.g.:
```
go func() {
    var wg sync.WaitGroup
   for i := 0; i < defaultMaxDataGetCoroutines; i++ {
		wg.Add(1)
		go getInputData(i, &wg, tasksChan, resultsChan, errorsChan)
	}
	wg.Wait()
	close(resultsChan)
	close(errorsChan)
}
```

Reply:
Good point, thank you!